### PR TITLE
Use RFC 3986-compliant URIs for file paths

### LINKED ISSUE #218: bug: Autocomplete and Chat Function Not Working in WebStorm
### Installation Information

WebStorm 2023.3.1
Build #WS-233.11799.293, built on December 11, 2023
Licensed to Helge Kosuch
Subscription is active until December 22, 2023.
Runtime version: 17.0.9+7-b1087.7 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 2048M
Cores: 8
Registry:
  documentation.show.toolbar=true
  ide.experimental.ui=true
Non-Bundled Plugins:
  Batch Scripts Support (1.0.13)
  com.sourcegraph.jetbrains (5.2.13806-nightly)

### Describe the bug

Summary:

Initial Report: Customer HelgeHelge reported that while the chat function in WebStorm was operational, the autocomplete feature was not working.

Troubleshooting Attempt: Justin Dorfman suggested restarting WebStorm, as he initially replicated the issue but found it resolved after a restart.

Customer Follow-Up: After restarting, HelgeHelge reported no change in the issue, stating that autocomplete was still non-functional.

Further Investigation: Justin inquired about the size of the codebase, to which HelgeHelge replied it was small, but noted there were many node modules.

Gitignore Check: Justin confirmed that the node modules were in a .gitignore file.

Additional Troubleshooting Steps: Justin advised creating a new file, adding some code, and trying the autocomplete command again, followed by taking a screenshot. He also requested the idea.log file for further analysis.

Customer's Last Update: HelgeHelge reported a persistent issue with a blinking cursor and expressed doubt that any change would occur.

### Expected behavior

* Autocomplete & Chat to work as expected

### Additional context (logs, images, etc)

```
2023-12-19 20:13:55,179 [ 506983]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-12-19 20:14:29,217 [ 541021]   INFO - #git4idea.commands.GitHandler - [Anca] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false add --ignore-errors -A -- dustin.ts
2023-12-19 20:14:31,023 [ 542827]   WARN - #c.i.j.n.i.l.NodeJsLocalInterpreter - Npm global package dir hasn't been calculated in 2000 ms, making a guess
2023-12-19 20:14:31,025 [ 542829]   INFO - #c.i.j.n.i.l.NodeJsLocalInterpreter - Guessed config prefix for D:\Program Files\nodejs\node.exe: C:\Users\helge\AppData\Roaming\npm
2023-12-19 20:14:31,045 [ 542849]   WARN - #com.sourcegraph.cody.agent.CodyAgent - No document found for file path file:///L:/Dropbox/o2/Nobref/Prog/JS/Anki/Anca/dustin.ts [
2023-12-19 20:14:31,045 [ 542849]   WARN - #com.sourcegraph.cody.agent.CodyAgent -   'file:///L:/Dropbox/o2/Nobref/Prog/JS/Anki/Anca/HelgeUtils.ts',
2023-12-19 20:14:31,046 [ 542850]   WARN - #com.sourcegraph.cody.agent.CodyAgent -   'file:///L:/Dropbox/o2/Nobref/Prog/JS/Anki/Anca/dustin.ts'
2023-12-19 20:14:31,046 [ 542850]   WARN - #com.sourcegraph.cody.agent.CodyAgent - ]
2023-12-19 20:14:31,395 [ 543199]   INFO - #c.i.p.m.i.ProjectDependenciesModelImpl - Count of declared packages: null, imported packages: 0
2023-12-19 20:14:31,398 [ 543202]   INFO - #c.i.p.m.i.ProjectDependenciesModelImpl - Count of declared packages: null, imported packages: 0
2023-12-19 20:14:32,395 [ 544199]   INFO - #c.i.p.m.ProjectDependenciesModuleAggregator - Supported models after update: [NpmProjectDependenciesModel]
2023-12-19 20:14:32,410 [ 544214]   INFO - #c.i.p.m.i.ProjectDependenciesModelImpl - Count of declared packages: 0, imported packages: 0
2023-12-19 20:14:33,342 [ 545146]   WARN - #c.i.j.n.i.l.NodeJsLocalInterpreter - Npm global package dir hasn't been calculated in 2000 ms, making a guess
2023-12-19 20:14:33,343 [ 545147]   INFO - #c.i.j.n.i.l.NodeJsLocalInterpreter - Guessed config prefix for D:\Program Files\nodejs\node.exe: C:\Users\helge\AppData\Roaming\npm
2023-12-19 20:14:33,640 [ 545444]   INFO - #c.i.j.n.i.l.NodeJsLocalInterpreter - "D:\Program Files\nodejs\npm.cmd" config get prefix=C:\Users\helge\AppData\Roaming\npm (4615 ms)
2023-12-19 20:14:46,010 [ 557814]   WARN - #com.sourcegraph.cody.agent.CodyAgent - recipe failed Error: Not implemented
2023-12-19 20:14:46,011 [ 557815]   WARN - #com.sourcegraph.cody.agent.CodyAgent -     at AgentEditor.showWarningMessage (C:\snapshot\dist\agent.js)
2023-12-19 20:14:46,012 [ 557816]   WARN - #com.sourcegraph.cody.agent.CodyAgent -     at ExplainCodeDetailed.getInteraction (C:\snapshot\dist\agent.js)
2023-12-19 20:14:46,012 [ 557816]   WARN - #com.sourcegraph.cody.agent.CodyAgent -     at Object.executeRecipe (C:\snapshot\dist\agent.js)
2023-12-19 20:14:46,012 [ 557816]   WARN - #com.sourcegraph.cody.agent.CodyAgent -     at C:\snapshot\dist\agent.js
2023-12-19 20:14:46,012 [ 557816]   WARN - #com.sourcegraph.cody.agent.CodyAgent -     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2023-12-19 20:15:06,487 [ 578291]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-12-19 20:15:32,748 [ 604552]   INFO - #c.i.o.v.n.p.FSRecords - Checking VFS started
2023-12-19 20:15:32,862 [ 604666]   INFO - #c.i.o.v.n.p.FSRecords - 1364 file records checked: 1346 children, 164 contents
2023-12-19 20:15:33,027 [ 604831]   INFO - #c.i.o.v.n.p.FSRecords - Checking VFS finished (healthy): VFSHealthCheckReport(recordsReport=FileRecordsReport(fileRecordsChecked=1364, fileRecordsDeleted=3, nullNameIds=0, unresolvableNameIds=0, notNullContentIds=164, unresolvableContentIds=0, unresolvableAttributesIds=0, nullParents=0, childrenChecked=1346, inconsistentParentChildRelationships=0, generalErrors=0), rootsReport=RootsReport(rootsCount=14, rootsWithParents=0, rootsDeletedButNotRemoved=0, generalErrors=0), namesEnumeratorReport=NamesEnumeratorReport(namesChecked=2181, namesResolvedToNull=0, idsResolvedToNull=0, inconsistentNames=0, generalErrors=0), contentEnumeratorReport=ContentEnumeratorReport(contentRecordsChecked=503, generalErrors=0), timeTaken=254.635900ms)
2023-12-19 20:15:33,266 [ 605070]   INFO - #c.i.i.s.p.i.SharedIndexMainZipStorage - Shared Indexes Storage is opened and empty
2023-12-19 20:15:33,275 [ 605079]   WARN - #c.i.u.x.Binding - no accessors for com.intellij.toolWindow.ToolWindowDefaultLayoutManager$ToolWindowLayoutStorageManagerState
2023-12-19 20:15:33,455 [ 605259]   WARN - #c.i.f.f.WSLInstallationsCollector - Failed to load installed WSL distributions: Failed to run C:\WINDOWS\system32\wsl.exe --list --verbose: {exitCode=-1, timeout=false, cancelled=false, stdout=Windows Subsystem for Linux has no installed distributions.

Use 'wsl.exe --list --online' to list available distributions
and 'wsl.exe --install <Distro>' to install.

Distributions can also be installed by visiting the Microsoft Store:
https://aka.ms/wslstore
Error code: Wsl/WSL_E_DEFAULT_DISTRO_NOT_FOUND
, stderr=}, done in 482 ms
2023-12-19 20:16:25,013 [ 656817]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-12-19 20:16:58,262 [ 690066]   INFO - #c.i.u.MemTester - not supported: Windows is not supported
2023-12-19 20:17:26,109 [ 717913]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-12-19 20:17:46,199 [ 738003]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-12-19 20:17:56,000 [ 747804]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-12-19 20:18:10,013 [ 761817]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
```
image

First thought it might be a Dropbox issue, but happen in a normal folder as well.
image

### PR DESCRIPTION
**Pull request by exigow**
Fixes #218

The Java URIs uses a different standard (RFC 2396) for encoding than the library vscode-uri used by Cody agent (RFC 3986).
Due to the fact that I was unable to find a reasonable library, this patch merely imitates the encoding and normalization of such addresses.

Additional change: Remove unused CodyAgentDocuments.

## Test Plan

1. Verify basic features on Windows
2. Verify basic features on Linux

### CODE REVIEW

**Comment by exigow** [comment]
Testing on Windows:
* chat works
* autocompletion works
* commands works

**Comment by exigow** [comment]
Testing on Linux:
* chat works
* autocompletion works
* command works

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.sourcegraph.cody.agent.protocol.util
+
+object Rfc3986UriEncoder {
+
+  // todo solve this with library
+  fun encode(uri: String): String {
+    val isWindowsPath = uri.matches("^file:///[A-Z]:/.*".toRegex())
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt]: does it make sense to extract `"^file:///[A-Z]:/.*".toRegex()` to a field? idk what `toRegex` does and if there is any optimization

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.sourcegraph.cody.agent.protocol.util
+
+object Rfc3986UriEncoder {
+
+  // todo solve this with library
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt:5]: have you found any libs that could do that?

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  fun encode(uri: String): String {
+    val isWindowsPath = uri.matches("^file:///[A-Z]:/.*".toRegex())
+    if (isWindowsPath) {
+      val found = "file:///([A-Z]):".toRegex().find(uri)!!
+      val partition = found.groups[1]!!.value
+      return uri.replace("file:///$partition:", "file:///${partition.lowercase()}%3A")
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt:11]: basically, what we want to achieve is to replace the colon symbol (the one that appears after the partition name) with a `%3A`. is that right?

File: src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.sourcegraph.cody.agent.protocol.util
+
+import junit.framework.TestCase
+
+class Rfc3986UriEncoderTest : TestCase() {
```
Comment (mkondratek) [inline src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt:5]: 💟

File: src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java [hunk 5]
```diff
@@ -18,7 +18,6 @@ public class CodyAgentClient {
 
   private static final Logger logger = Logger.getInstance(CodyAgentClient.class);
   @Nullable public CodyAgentServer server;
-  @Nullable public CodyAgentDocuments documents;
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java:21]: this change is unrelated, right? is this field unused?

**Review by mkondratek — APPROVED** [review]
I tested it on macOS:
- autocompletes (a few, with a cycle over suggestion)
- chat messages

works 🚀

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt (hunk 2 above)
Comment (exigow) [reply src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt:5]: Yeah, that's why I've added `todo`. Libs I've checked so far:
* Apache HTTP client encoder - comma not encoded, partition not normalized
* Spring Web - similar, but also version I've checked has some CVEs
* https://github.com/dmfs/uri-toolkit - I haven't checked that, but development is not active

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt (hunk 1 above)
Comment (exigow) [reply src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt]: Possibly. From what I've seen, the matcher is statefu but the compiled Regex itself might be held as a static field. But I don't think we should worry about that.

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt (hunk 3 above)
Comment (exigow) [reply src/main/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoder.kt:11]: Colon symbol and partition letter as lowercase for Windows paths - but possibly more.

BTW encoded colon with uppercase partition letter it's not working.

File: src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java (hunk 5 above)
Comment (exigow) [reply src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java:21]: Correct, unrelated, but I've created static factory method for TextDocument and it's turned out that most of usages are dead code. So i removed that (see: CodyAgentDocuments).

File: src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+  fun `test encode Windows path`() {
+    val fixedUri = Rfc3986UriEncoder.encode("file:///C:/Users/user/Test.java")
```
Comment (pkukielka) [inline src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt:8]: Can you please add test for `"file:///c:/Users/user/Test.java"`?

Comment (exigow) [reply src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt:8]: Sure!

Comment (exigow) [reply src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt:8]: BTW can you elaborate @pkukielka? It seems to me that partitions are always capitalized.

Comment (pkukielka) [reply src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt:8]: Well, maybe it's defensive programming, but windows theoretically is case insensitive, so what guarantees you that you won't get lower letter there somehow?
I understand that we do not expect it to happen :)

Comment (exigow) [reply src/test/kotlin/com/sourcegraph/cody/agent/protocol/util/Rfc3986UriEncoderTest.kt:8]: Fixed

**Review by pkukielka — APPROVED** [review]
LGTM

**Comment by exigow** [comment]
Tested IDEs on Windows (incl. autocompletion, chat and commands):
* IntelliJ IDEA 2022.1 CE (via runIde)
* IntelliJ IDEA 2023.3.2 UE (manual installation)
* WebStorm 2023.3.1 CE (manual installation)

**Pull request merged**